src/Orleans.Runtime/Placement/ResourceOptimizedPlacementDirector.cs

```diff
@@ -0,0 +1,257 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using Orleans.Runtime.Configuration.Options;
+
+namespace Orleans.Runtime.Placement;
+
+internal sealed class ResourceOptimizedPlacementDirector : IPlacementDirector, ISiloStatisticsChangeListener
+{
+    readonly record struct ResourceStatistics(float? CpuUsage, float? AvailableMemory, long? MemoryUsage, long? TotalPhysicalMemory, bool IsOverloaded);
+
+    Task<SiloAddress> _cachedLocalSilo;
+
+    readonly ResourceOptimizedPlacementOptions _options;
+    readonly ConcurrentDictionary<SiloAddress, ResourceStatistics> siloStatistics = [];
+
+    readonly DualModeKalmanFilter<float> _cpuUsageFilter = new();
+    readonly DualModeKalmanFilter<float> _availableMemoryFilter = new();
+    readonly DualModeKalmanFilter<long> _memoryUsageFilter = new();
+
+    /// <summary>
+    /// 1 / (1024 * 1024)
+    /// </summary>
+    const float physicalMemoryScalingFactor = 0.00000095367431640625f;
```

Comment:
```suggestion
    /// <summary>
    /// 1 / (1024 * 1024)
    /// </summary>
    private const float PhysicalMemoryScalingFactor = 0.00000095367431640625f;
    
    private readonly ResourceOptimizedPlacementOptions _options;
    private readonly ConcurrentDictionary<SiloAddress, ResourceStatistics> siloStatistics = [];

    private readonly DualModeKalmanFilter<float> _cpuUsageFilter = new();
    private readonly DualModeKalmanFilter<float> _availableMemoryFilter = new();
    private readonly DualModeKalmanFilter<long> _memoryUsageFilter = new();

    private Task<SiloAddress> _cachedLocalSilo;
```

* All fields have access qualifiers.
* Ordered: const, readonly, mutable.